### ISSUE TITLE
connectOrCreate fails during upsert

### ISSUE THREAD

**Issue by jazerix** [prisma/prisma#3612]
## Bug description

@kazimirzak and I have run into some issues while trying to upsert and connectOrCreate at the same time.

When trying to use connectOrCreate during an `upsert()`, prisma will throw the following error:
> Error occurred during query execution:
EnvVarNotFound("Expected parent binding \'4\' to be present.")
    at PrismaClientFetcher.request (...\node_modules\@prisma\client\runtime\index.js:1:228459)

"4" is not part of any variable being passed to the upsert function.

The data we're trying create looks as follows:

```typescript
this.client.serverToSinner.upsert({
    where: {
        serverId_sinnerId: {
            serverId,
            sinnerId
        }
    },
    create: {
        Server: {
            connect: {
                id: serverId
            }
        },
        Sinner: {
            connectOrCreate: {
                where: {
                    id: sinnerId
                },
                create: {
                    id: sinnerId,
                    username: sinnerUsername,
                }
            }
        },
        absolvedAt
    },
    update: {
        absolvedAt
    }
}
```

`serverToSinner` is an intermediate table.

It's worth noting that, if we change `connectOrCreate` to `connect` and make sure the record is already there, it works perfectly fine. 
## How to reproduce

1. Create 2 tables - the primary key on each being a string (varchar in our case)
2. Create an intermediate table using a composite primary key from the two tables created in 1. 
3. Add a foreign key on the intermediate table connecting _primaryKey1_ to _table1_ and _primaryKey2_ to _table2_
4. On _table1_ create a record.
5. Create a call to `upsert(..)` where you connect to the record on _table1_ and then `connectOrCreate` on _table2_, thus a new record should be created on _table2_ and then a record connecting _record1_ and _record2_ in the intermediate table.

The database should look somewhat like this:
![image](https://user-images.githubusercontent.com/4068103/92983261-8915ad80-f4a2-11ea-8da4-bca299f203d9.png)

## Expected behavior

1. Prisma should realize that the intermediate relationship doesn't exist during `upsert` ✔️ 
2. Prisma should first check _table1_ for an appropriate record.
3. Prisma then checks _table2_ for a second record.
4. Upon realizing that the record doesn't exist on _table2_ Prisma should create the record.
5. The intermediate table should be updated, connecting the newly created record on _table2_ to the record on _table1_.

## Prisma information

```
model Server {
  id             String           @id
  name           String?
  channelId      String?
  ServerToSinner ServerToSinner[] @relation("ServerTo_ServerToSinner")
}

model Sinner {
  id             String           @id
  username       String
  ServerToSinner ServerToSinner[] @relation("SinnerTo_ServerToSinner")
}

model ServerToSinner {
  serverId       String
  sinnerId       String
  questionsAsked Int      @default(0)
  absolvedAt     DateTime
  Server         Server   @relation("ServerTo_ServerToSinner", fields: [serverId], references: [id])
  Sinner         Sinner   @relation("SinnerTo_ServerToSinner", fields: [sinnerId], references: [id])

  @@id([serverId, sinnerId])
  @@index([sinnerId], name: "fk_sinner")
  @@map("_ServerToSinner")
}
```

## Environment & setup

```
OS: Windows 10, Version 2004
Database: MySQL
Node.js version: 14.9
Primsa: 2.6.2
```

**Comment by stefanoruth** [comment]
Can confirm and reproduce this issue, gets the same error `Expected parent binding '4' to be present.` when running the query.

One thing to note tho is that it works if you turn it around and and use connectOrCreate on Server and just connect on Sinner then it just works.

If you then use connectOrCreate on both queries then error changes binding number to  `Expected parent binding '8' to be present.` where it looks like it run the Server part of the query find but then fails when it comes to Sinner, making it seems like there is some issue with the relation to Sinner.

Also trying moving the @id of the ServerToSinner table to its own column and making that its only index but its changes nothing, unfortunately.

**Comment by pantharshit00** [comment]
I can confirm this issue.

Here is a minimal reproduction in a form of a quick script that we can while fixing this: https://github.com/harshit-test-org/prisma-issue-3612

**Comment by pantharshit00** [comment]
This might be a duplicate of https://github.com/prisma/prisma-client-js/issues/764

**Comment by dpetrick** [comment]
Confirmed duplicate of https://github.com/prisma/prisma-client-js/issues/764

**Comment by dpetrick** [comment]
A fix has been merged, this will be available in the 2.10 release.
